DeeperFunc/Maybe.swift

```diff
@@ -0,0 +1,89 @@
+//
+//  Maybe.swift
+//  DeeperFunc
+//
+//  Created by Ilya Puchka on 29/10/2017.
+//  Copyright © 2017 Ilya Puchka. All rights reserved.
+//
+
+import Foundation
+
+func maybe(_ route: RoutePattern<Void, Path>) -> RoutePattern<Void, Path> {
+    return .init(parse: { route.parse($0) ?? ($0, ()) }, print: route.print, template: "(\(route.template))")
+}
+
+func maybe<A, S>(_ route: RoutePattern<A, S>) -> RoutePattern<A?, S> {
+    return .init(parse: { url in
+        guard let result = route.parse(url) else { return (url, nil) }
+        return (result.rest, .some(result.match))
+    }, print: {
+        return $0.flatMap(route.print)
+    }, template: "(\(route.template))")
+}
+
+infix operator /? : MultiplicationPrecedence
+
+extension RoutePattern where S == Path {
+    
+    public static func /?(lhs: RoutePattern, rhs: RoutePattern<Void, S>) -> RoutePattern {
+        let rhs = maybe(rhs)
+        return .init(parse: parseLeft(lhs, rhs), print: printLeft(lhs, rhs), template: templateAnd(lhs, rhs))
+    }
+    
+    public static func /?(lhs: RoutePattern<Void, S>, rhs: RoutePattern) -> RoutePattern<A?, S> {
+        let rhs = maybe(rhs)
+        return .init(parse: parseRight(lhs, rhs), print: printRight(lhs, rhs), template: templateAnd(lhs, rhs))
+    }
+    
+    public static func /?<B>(lhs: RoutePattern, rhs: RoutePattern<B, S>) -> RoutePattern<(A, B?), S> {
```

Comment:
maybe there should be `/?` and `>/?` so that it expresses that parameter from left part is carried on or not? but `>/?` feels like set of random symbols